extensions/native/recursion/src/fri/two_adic_pcs.rs

```diff
@@ -399,73 +365,146 @@ where
+        for i in 0..MAX_LOG_WIDTH {
+            builder.set(&pow_of_alpha, i, current);
+            builder.assign(&current, current * current);
+        }
     }
-    let round_alpha_pows: Array<C, Array<C, Ext<_, _>>> = builder.array(rounds.len());
-    iter_zip!(builder, rounds, round_alpha_pows).for_each(|ptr_vec, builder| {
-        let round = builder.iter_ptr_get(&rounds, ptr_vec[0]);
+
+    iter_zip!(builder, rounds, ret).for_each(|ptr_vec, builder| {
+        let round = builder.iter_ptr_get(rounds, ptr_vec[0]);
+        let permutation = round.permutation;
+        let to_perm_index = |builder: &mut Builder<_>, k: RVar<_>| {
+            // Always no permutation in static mode
+            if builder.flags.static_only {
+                builder.eval(k)
+            } else {
+                let ret: Usize<_> = builder.uninit();
+                builder.if_eq(permutation.len(), RVar::zero()).then_or_else(
+                    |builder| {
+                        builder.assign(&ret, k);
+                    },
+                    |builder| {
+                        let value = builder.get(&permutation, k);
+                        builder.assign(&ret, value);
+                    },
+                );
+                ret
+            }
+        };
+
+        let ov_buff: Array<C, Array<C, Felt<C::F>>> = builder.array(round.mats.len());
+        let perm_ov_buff: Array<C, Array<C, Felt<C::F>>> = builder.array(round.mats.len());
+        let batch_dims: Array<C, DimensionsVariable<C>> = builder.array(round.mats.len());
         let mat_alpha_pows: Array<C, Ext<_, _>> = builder.array(round.mats.len());
-        iter_zip!(builder, round.mats, mat_alpha_pows).for_each(|ptr_vec, builder| {
-            assert!(!builder.flags.static_only);
-            let mat = builder.iter_ptr_get(&round.mats, ptr_vec[0]);
-            let local = builder.get(&mat.values, 0);
-            let mat_alpha_pow: Ext<_, _> = {
-                let width = local.len().get_var();
-                // This is dynamic only so safe to cast.
-                let width_f = builder.unsafe_cast_var_to_felt(width);
-                let bits = builder.num2bits_f(width_f, MAX_LOG_WIDTH as u32);
-                let ret: Ext<_, _> = builder.eval(C::EF::ONE.cons());
-                for i in 0..MAX_LOG_WIDTH {
-                    let bit = builder.get(&bits, i);
-                    builder.if_eq(bit, RVar::one()).then(|builder| {
-                        let to_mul = builder.get(&pow_of_alpha, i);
-                        builder.assign(&ret, ret * to_mul);
-                    });
+        let log_batch_max_height: Usize<_> = {
+            let log_batch_max_index = to_perm_index(builder, RVar::zero());
+            let mat = builder.get(&round.mats, log_batch_max_index);
+            let domain = mat.domain;
+            builder.eval(domain.log_n + RVar::from(log_blowup))
+        };
+
+        builder
+            .range(0, round.mats.len())
```

Comment:
I think it's more efficient to use `iter_zip!` over the different arrays you need to `iter_set, iter_get`